Images auto update

presztak opened the pull request:
It's a third part of "Add `images.*` and `backups.*` server config keys to per-project #7900" issue.

Not sure about projectsAutoUpdateTime map but we need to keep somewhere times of autoUpdateTask execution per project. So we know when run it next time.
I've made also some cleanup staff related with remote_cache_expiry and auto_update_interval keys.

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
Instead of storing a timestamp per project which seems quite brittle especially when considering updates/addition/deletion of projects within a cluster, maybe a solution would be to add a `startTime` property to `Daemon` and then determine what project need updating based on the project update interval and the daemon startTime?

stgraber commented:
@presztak does that sound like a viable solution to you?

presztak commented:
@stgraber I thought about that kind of solution but have one doubt. For example if we have project with interval equal 3 and after first update someone change interval to value 2 then next update will be in the 4th hour after daemon start, so one hour after first update (because we have only daemon startTime so we run update task when (elapsedHours % interval == 0), right?). If this is acceptable we can implement that in this way.

stgraber commented:
Yeah, that's fine with me. It's effectively similar to if you changed a cron pattern from `*/3` to `*/4`, your next run may happen sooner but the interval will be correct after that.

presztak commented:
@stgraber is current solution ok for you?

stgraber commented on lxd/images.go:
```diff
@@ -1111,19 +1112,7 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
-		return interval, nil
+		return time.Hour, nil
 	}
 	return f, schedule
```
`return f, task.Every(time.Hour)` should be a shorter equivalent here.

stgraber commented:
Looks good to me, just one tiny optimization that can be done to shed a few lines of code and this will be good to go!

stgraber commented:
jenkins: test this please

stgraber commented:
Hmm, looks like this is breaking the expiry test.

presztak commented:
@stgraber Yes. This is caused by removing this lines:
```
if !d.os.MockMode {
d.taskPruneImages.Reset()
}
```
from lxd/api_1.0.go.

So now after changing images.remote_cache_expiry value, task is not fired. And in `test_image_expiry` we have:
```
# Trigger the expiry
  lxc_remote config set l2: images.remote_cache_expiry 1
```

How we want to solve this? I removed this lines because I thought if we have fixed interval so we don't need to reset (fire) task after changing interval. Should we revert that code? Or maybe fix test but question is how fire task in test?

presztak commented:
@stgraber oh, I see that interval value for `pruneExpiredImagesTask` was always constant, and even then `pruneExpiredImagesTask` was reset after changing images.remote_cache_expiry value. So I think the best solution will be revert that part of code which reset `pruneExpiredImagesTask`. Is it ok for you?

stgraber commented:
yeah, I think that should be fine

presztak commented:
@stgraber should be better right now.

stgraber commented:
jenkins: test this please

tomponline commented:
jenkins: test this please

stgraber commented:
Thanks!
